### ISSUE TITLE
Bottleneck phase produces bad files (invalid or with poor accuracy results) on some configuration

### ISSUE THREAD

**Issue by francois-dorin** [dotnet/machinelearning#4703]
### System information

- **OS version/distro**: Windows 10 x64
- **.NET Version (eg., dotnet --info)**: 3.1.101
- **ML.NET Version**: 1.4.0 
- **CPU : AMD Ryzen 9 3900X 12-Core**

### Issue

- **What did you do?**
Running a simple example of image classification, based on https://docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/image-classification-api-transfer-learning

- **What happened?**
Using only CPU, the process completes but provides a very poor accuracy (0.014)
Using GPU, the process passes the bottleneck phase but cannot begin the train due to an error.

- **What did you expect?**
Using only CPU, I expect a better result (around 0.70).
Using GPU, I expect the process to continue and train.

- **Additionnal informations**
The same program with the same dataset works fine on another computer (Intel based CPU) and have a correct accuracy (0.70).
No problem occurs when using the GPU.
If I copy/paste the cached files from the working computer to my computer, the process is working fine (with or without GPU) and have the expected accuracy (0.70).
So, I guess the problem relies in the files generated during the bottleneck phase and is related to the hardware configuration, since same program with same dataset have very different behaviours.

### Source code / logs
Complete logs (CPU Only) when using generated cached files during the bottleneck phase
```
*** Training the image classification model with DNN Transfer Learning on top of the selected pre-trained model/architecture ***
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel started
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel finished. Elapsed 00:00:00.0083464.
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel disposed
2020-01-24 16:03:09.939754: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
Saver not created because there are no variables in the graph to restore
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:       0,01 Epoch:   0, Accuracy:  0,6121079, Cross-Entropy: 4,300151E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   0, Accuracy: 0,01333781
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:       0,01 Epoch:   1, Accuracy:  0,6213207, Cross-Entropy: 3,970512E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   1, Accuracy: 0,01346435
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:     0,0094 Epoch:   2, Accuracy:  0,6239617, Cross-Entropy: 3,621145E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   2, Accuracy: 0,01293448
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:     0,0094 Epoch:   3, Accuracy:  0,6224542, Cross-Entropy: 3,632903E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   3, Accuracy: 0,01363834
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:   0,008836 Epoch:   4, Accuracy:  0,6240405, Cross-Entropy: 3,406656E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   4, Accuracy: 0,01359484
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate:   0,008836 Epoch:   5, Accuracy:  0,6223006, Cross-Entropy: 3,40876E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   5, Accuracy: 0,01343272
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,008305839 Epoch:   6, Accuracy:  0,6249708, Cross-Entropy: 3,204811E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   6, Accuracy:  0,0134169
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,008305839 Epoch:   7, Accuracy:  0,6249651, Cross-Entropy: 3,184228E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   7, Accuracy: 0,01392107
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,007807489 Epoch:   8, Accuracy:  0,6232149, Cross-Entropy: 3,004976E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   8, Accuracy: 0,01329432
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,007807489 Epoch:   9, Accuracy:  0,6227242, Cross-Entropy: 3,030696E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:   9, Accuracy: 0,01335956
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,00733904 Epoch:  10, Accuracy:  0,6246268, Cross-Entropy: 2,813951E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  10, Accuracy: 0,01338527
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,00733904 Epoch:  11, Accuracy:  0,6228801, Cross-Entropy: 2,828588E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  11, Accuracy: 0,01380046
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006898697 Epoch:  12, Accuracy:  0,6257862, Cross-Entropy: 2,645051E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  12, Accuracy: 0,01335957
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006898697 Epoch:  13, Accuracy:  0,6236802, Cross-Entropy: 2,659039E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  13, Accuracy: 0,01352564
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006484775 Epoch:  14, Accuracy:  0,6229582, Cross-Entropy: 2,493187E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  14, Accuracy: 0,01385187
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006484775 Epoch:  15, Accuracy:  0,6234171, Cross-Entropy: 2,496486E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  15, Accuracy: 0,01407331
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006095689 Epoch:  16, Accuracy:  0,6221284, Cross-Entropy: 2,340333E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  16, Accuracy: 0,01392898
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,006095689 Epoch:  17, Accuracy:  0,6231769, Cross-Entropy: 2,359828E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  17, Accuracy: 0,01311638
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,005729948 Epoch:  18, Accuracy:  0,6233777, Cross-Entropy: 2,180089E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  18, Accuracy: 0,01357309
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,005729948 Epoch:  19, Accuracy:  0,6225668, Cross-Entropy: 2,219908E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  19, Accuracy: 0,01381233
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 0,005386151 Epoch:  20, Accuracy:  0,6234834, Cross-Entropy: 2,070789E+12
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch:  20, Accuracy: 0,01392503
Saver not created because there are no variables in the graph to restore
```

End logs (CPU Only) when using cached files from a working computer (more epoch and better results)
```Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch: 222, Accuracy:  0,7872852
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 1,040379E-05 Epoch: 223, Accuracy:  0,8063375, Cross-Entropy:  0,7515593
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch: 223, Accuracy:  0,7873719
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 9,779563E-06 Epoch: 224, Accuracy:  0,8058808, Cross-Entropy:  0,7523516
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch: 224, Accuracy:  0,7878902
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 9,779563E-06 Epoch: 225, Accuracy:  0,8060154, Cross-Entropy:  0,7517532
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch: 225, Accuracy:  0,7882423
Phase: Training, Dataset used:      Train, Batch Processed Count: 9177, Learning Rate: 9,19279E-06 Epoch: 226, Accuracy:  0,8054689, Cross-Entropy:  0,7526353
Phase: Training, Dataset used: Validation, Batch Processed Count: 2299, Epoch: 226, Accuracy:   0,788062
```

Logs when using GPU
```
*** Training the image classification model with DNN Transfer Learning on top of the selected pre-trained model/architecture ***
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel started
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel finished. Elapsed 00:00:00.0088906.
[Source=ImageClassificationTrainer; Ensuring meta files are present., Kind=Trace] Channel disposed
2020-01-24 16:11:47.949314: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2020-01-24 16:11:47.960130: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library nvcuda.dll
2020-01-24 16:11:47.998004: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties:
name: GeForce RTX 2070 SUPER major: 7 minor: 5 memoryClockRate(GHz): 1.785
pciBusID: 0000:2d:00.0
2020-01-24 16:11:48.000866: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2020-01-24 16:11:48.003049: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-01-24 16:11:49.018802: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-01-24 16:11:49.020743: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0
2020-01-24 16:11:49.021869: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N
2020-01-24 16:11:49.023948: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6290 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070 SUPER, pci bus id: 0000:2d:00.0, compute capability: 7.5)
Saver not created because there are no variables in the graph to restore
2020-01-24 16:11:49.891634: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties:
name: GeForce RTX 2070 SUPER major: 7 minor: 5 memoryClockRate(GHz): 1.785
pciBusID: 0000:2d:00.0
2020-01-24 16:11:49.894813: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2020-01-24 16:11:49.897236: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-01-24 16:11:49.899756: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-01-24 16:11:49.902237: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0
2020-01-24 16:11:49.903756: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N
2020-01-24 16:11:49.906049: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6290 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070 SUPER, pci bus id: 0000:2d:00.0, compute capability: 7.5)
Can't copy 160160 bytes of a tensor into another with 80080 bytes buffer.
         [[{{node _arg_input_1/BottleneckInputPlaceholder_0_0}}]]
```

**Comment by frank-dong-ms** [comment]
@francois-dorin could you please provide a repro project and corresponding dataset so we can look further into this?

**Comment by mstfbl** [comment]
Hi @francois-dorin ,

I tested the image classification sample you've linked on an Azure VM with an AMD EPYC 7452 32-Core Processor. In these results I did not obtain the low accuracy results you've provided. I've attached logs of running this on Intel and AMD CPUs  to this comment, please feel free to take a look.

I'd suggest cleaning your build, and rebuilding to see if your results improve. Also please make sure your GPU libraries are up to date. If you have any further questions, please feel free to share. Thanks!

[image_classification_amd.txt](https://github.com/dotnet/machinelearning/files/4685867/image_classification_amd.txt)
[image_classification_normal.txt](https://github.com/dotnet/machinelearning/files/4685868/image_classification_normal.txt)

**Comment by pantokrr** [comment]
Hi!
I have the exact same issue, ending in:
```
Tensorflow.TensorflowException: 'Can't copy 80080 bytes of a tensor into another with 40040 bytes buffer.
	 [[{{node _arg_input_1/BottleneckInputPlaceholder_0_0}}]]'
```

**Comment by mstfbl** [comment]
Hi @pantokrr and @francois-dorin ,

The `TensorflowException` you've been having has been fixed in Issue #4708 , which is available in our newest release of ML .NET v1.5.0. Please feel free to download its NuGet package from [here](https://www.nuget.org/packages/Microsoft.ML/), and let us know if your issue persists. Thanks!

**Comment by francois-dorin** [comment]
Hi @frank-dong-ms and @mstfbl 
Sorry for the delay to answer but a lot of work currently.

I did not try with the latest release of ML.NET via Nuget, but using a more recent developpement version from git repo solved the problem. So I guess the bug was already fixed.

Thank you guys.
